In [4]:
import cobra 
from thermo_flux.io import load_excel as ex
import thermo_flux
import pandas as pd

In [2]:
INPUT_MODEL = "datafiles/model.xlsx"
INPUT_KEGGS = "datafiles/ecoli_kegg_id.csv"

In [6]:
condition = 'GLC'

## load stoichiometric model 
and add same reactions than for 1_tfva

In [3]:
tmodel = ex.create_thermo_model('ecoli', model_excel=INPUT_MODEL, keggids_csv=INPUT_KEGGS, edit_mets={})
# ADD HYDROXYBENZOATE TRANSPORT & EXCHANGE:
# Define extracellular orotate:
hbz_e = cobra.Metabolite(id="4hbz_e", compartment="e")
hbz_e = thermo_flux.core.metabolite.ThermoMetabolite(hbz_e, model=tmodel)
hbz_e.annotation = tmodel.metabolites.get_by_id("4hbz_c").annotation  
tmodel.metabolites.append(hbz_e)

# Hydroxybenzoate H+ antiporter (aaeB):
HBZt3 = cobra.Reaction("HBZt3")
HBZt3.lower_bound = -1000
HBZt3.upper_bound = 1000
HBZt3.add_metabolites({tmodel.metabolites.get_by_id("4hbz_e"): +1,
                    tmodel.metabolites.h_e: -1,
                    tmodel.metabolites.get_by_id("4hbz_c"): -1,
                    tmodel.metabolites.h_c: +1})
HBZt3 = thermo_flux.core.reaction.ThermoReaction(HBZt3, model=tmodel)

# Exchange:
EX_4hbz = cobra.Reaction("EX_4hbz")
EX_4hbz.add_metabolites({tmodel.metabolites.get_by_id("4hbz_e"): -1})
EX_4hbz = thermo_flux.core.reaction.ThermoReaction(EX_4hbz, model=tmodel)

tmodel.add_reactions([HBZt3, EX_4hbz])
# ADD ISOPROPYLMALATE TRANSPORT & EXCHANGE:
# Define extracellular orotate:
ipm_e = cobra.Metabolite(id="3c3hmp_e", compartment="e")
ipm_e = thermo_flux.core.metabolite.ThermoMetabolite(ipm_e, model=tmodel)
ipm_e.annotation = tmodel.metabolites.get_by_id("3c3hmp_c").annotation  
tmodel.metabolites.append(ipm_e)

# Diffusion across the membrane:
ipm_diff = cobra.Reaction("IPMex")
ipm_diff.lower_bound = -1000
ipm_diff.upper_bound = 1000
ipm_diff.add_metabolites({tmodel.metabolites.get_by_id("3c3hmp_c"): -1,
                        tmodel.metabolites.get_by_id("3c3hmp_e"): +1})
ipm_diff = thermo_flux.core.reaction.ThermoReaction(ipm_diff, model=tmodel)

# Exchange:
EX_3c3hmp = cobra.Reaction("EX_3c3hmp")
EX_3c3hmp.add_metabolites({tmodel.metabolites.get_by_id("3c3hmp_e"): -1})
EX_3c3hmp = thermo_flux.core.reaction.ThermoReaction(EX_3c3hmp, model=tmodel)

tmodel.add_reactions([ipm_diff, EX_3c3hmp])

# ADD OROTATE TRANSPORT AND EXCHANGE
# Define extracellular orotate:
oro_e = cobra.Metabolite(id="orot_e", compartment="e")
oro_e = thermo_flux.core.metabolite.ThermoMetabolite(oro_e, model=tmodel)
oro_e.annotation = tmodel.metabolites.orot_c.annotation  
tmodel.metabolites.append(oro_e)

# Diffusion across the membrane:
oro_diff = cobra.Reaction("OROTex")
oro_diff.lower_bound = -1000
oro_diff.upper_bound = 1000
oro_diff.add_metabolites({tmodel.metabolites.orot_c: -1,
                        tmodel.metabolites.orot_e: +1})
oro_diff = thermo_flux.core.reaction.ThermoReaction(oro_diff, model=tmodel)

# Dicarboxylate/H+ symporter (dctA):
dcta = cobra.Reaction("DCTA")
dcta.lower_bound = -1000
dcta.upper_bound = 0
dcta.add_metabolites({tmodel.metabolites.orot_e: +1,
                    tmodel.metabolites.h_e: +1,
                    tmodel.metabolites.orot_c: -1,
                    tmodel.metabolites.h_c: -1})
dcta = thermo_flux.core.reaction.ThermoReaction(dcta, model=tmodel)

# Exchange:
EX_oro = cobra.Reaction("EX_oro")
EX_oro.add_metabolites({tmodel.metabolites.orot_e: -1})
EX_oro = thermo_flux.core.reaction.ThermoReaction(EX_oro, model=tmodel)

tmodel.add_reactions([oro_diff, dcta, EX_oro])

Set parameter Username
Set parameter LicenseID to value 2749904
Academic license - for non-commercial use only - expires 2026-12-04
['Parameters', 'Exchange reactions', 'Reactions', 'Biomass Composition', 'Transmembrane reactions', 'Metabolites', 'references', 'Transmembrane_reactions_reed', 'Transmembrane reactions_Orth', 'Transmembrane reactions old', 'Sheet3', 'log', 'subsystems']
*** Reading data from Reactions ***
unknown metabolite '2dhglcn[c]' created
unknown metabolite 'nadh[c]' created
unknown metabolite 'glcn[c]' created
unknown metabolite 'nad[c]' created
unknown metabolite 'nadph[c]' created
unknown metabolite 'nadp[c]' created
unknown metabolite '2dhguln[c]' created
unknown metabolite 'idon-L[c]' created
unknown metabolite '3hcinnm[c]' created
unknown metabolite 'o2[c]' created
unknown metabolite 'dhcinnm[c]' created
unknown metabolite 'h2o[c]' created
unknown metabolite '3hpppn[c]' created
unknown metabolite 'dhpppn[c]' created
unknown metabolite 'phthr[c]' created
unknow

Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`


cbi_c NOTHING DONE!
cbl1_c NOTHING DONE!
cdpdag1_c NOTHING DONE!
cmp_c NOTHING DONE!
paEC_c NOTHING DONE!
4c2me_c NOTHING DONE!
2p4c2me_c NOTHING DONE!
pphn_c NOTHING DONE!
3psme_c NOTHING DONE!
4hbz_c NOTHING DONE!
cinnm_c NOTHING DONE!
cenchddd_c NOTHING DONE!
pgEC_c NOTHING DONE!
clpnEC_c NOTHING DONE!
csn_c NOTHING DONE!
cpppg3_c NOTHING DONE!
pppg9_c NOTHING DONE!
bbtcoa_c NOTHING DONE!
crn_c NOTHING DONE!
crncoa_c NOTHING DONE!
gbbtn_c NOTHING DONE!
ctbtcoa_c NOTHING DONE!
ctbt_c NOTHING DONE!
ura_c NOTHING DONE!
utp_c NOTHING DONE!
ctp_c NOTHING DONE!
cyan_c NOTHING DONE!
tsul_c NOTHING DONE!
so3_c NOTHING DONE!
tcynt_c NOTHING DONE!
cynt_c NOTHING DONE!
acser_c NOTHING DONE!
h2s_c NOTHING DONE!
cyst-L_c NOTHING DONE!
cytd_c NOTHING DONE!
uri_c NOTHING DONE!
cdp_c NOTHING DONE!
dcmp_c NOTHING DONE!
dcdp_c NOTHING DONE!
fad_c NOTHING DONE!
fadh2_c NOTHING DONE!
dad-2_c NOTHING DONE!
din_c NOTHING DONE!
damp_c NOTHING DONE!
dadp_c NOTHING DONE!
12dgrEC_c NOTHING DONE!
26dap-M_c NO

## get reaction directions from TFVA

In [32]:
df_vbounds=pd.read_csv(f'ecoli_{condition}_vbounds.csv',sep='\t',index_col=0) ##sep might change in the future
df_vbounds
df_vbounds[abs(df_vbounds)<1e-6]=0


## get pre defined reaction directions 
look into bigg  for the initial model that was used to build the excel file (see supplementary note 2 Niebel 2019)


In [11]:
ecolimodel_bigg=cobra.io.load_model('iJR904')

Output()

In [48]:
## irr rxns in model from bIGG
# irr_rxns=[r.id for r in ecolimodel_bigg.reactions if r.bounds!=(-999999.0,999999.0)]
## how much of the irr rxns are in the tfva results df
common_rxns=[i.id for i in ecolimodel_bigg.reactions if i.id in df_vbounds.index]
init_bounds_df=pd.DataFrame(index=common_rxns,columns=['lb','ub'])
for rid in common_rxns:
    init_bounds_df.loc[rid]=ecolimodel_bigg.reactions.get_by_id(rid).bounds

In [49]:
init_bounds_df

,lb,ub
ALDD2x,0.0,999999.0
ABTA,0.0,999999.0
AMALT1,0.0,999999.0
AMALT2,0.0,999999.0
AMALT3,0.0,999999.0
...,...,...
SDPDS,0.0,999999.0
SDPTA,-999999.0,999999.0
SHK3Dr,-999999.0,999999.0
SHKK,0.0,999999.0


### now in the list of reactions that are  a priori irreversible, compare their bounds with tfva result

In [50]:
df_compare=pd.DataFrame(index=common_rxns)

def dir_from_bounds(row):
    if row['lb']>=0 :
        res='Forward'
    elif row['ub']<=0:
        res='Backward'
    else:
        res='Reversible'
    return res
            
df_compare['tfva_dir']=df_vbounds.loc[common_rxns].apply(dir_from_bounds,axis=1)
df_compare['init_dir']=init_bounds_df.apply(dir_from_bounds,axis=1)

In [51]:
mismatch = df_compare[df_compare['tfva_dir'] != df_compare['init_dir']]
mismatch.groupby(['tfva_dir', 'init_dir']).size()


tfva_dir    init_dir  
Backward    Forward         4
            Reversible     20
Forward     Reversible     15
Reversible  Forward       145
dtype: int64

### GLC : 
out of 415 common reactions (that are both in our model and the initial one from the BiGG database):
- 4 incorrectly labeled (opposite direction between tfva and predefined one ), 
- 35 that are more constrained with TFVA 
- 145 less constrained with TFVA